opslevel/provider.go

```diff
@@ -1,137 +1,278 @@
+		} else {
+			resp.Diagnostics.AddError(
+				"Missing OPSLEVEL_API_TOKEN",
+				"An OPSLEVEL_API_TOKEN is needed to authenticate with the opslevel client. "+
+					"This can be set as an environment variable or in the provider configuration block as 'api_token'.",
+			)
+		}
+	}
+	tflog.Debug(ctx, "opslevel client API token is set")
+
+	tflog.Debug(ctx, "Setting opslevel client API endpoint URL...")
+	if data.ApiUrl.IsNull() {
+		if apiUrl, ok := os.LookupEnv("OPSLEVEL_API_URL"); ok {
+			data.ApiUrl = types.StringValue(apiUrl)
+		} else {
+			data.ApiUrl = types.StringValue(defaultApiUrl)
+		}
+	}
+	tflog.Debug(ctx, "opslevel client API endpoint URL is set")
+
+	tflog.Debug(ctx, "Setting opslevel client API timeout...")
+	if data.ApiTimeout.IsNull() {
+		data.ApiTimeout = types.Int64Value(int64(defaultApiTimeout))
+
+		if apiTimeout, ok := os.LookupEnv("OPSLEVEL_API_TIMEOUT"); ok {
+			if timeout, err := strconv.Atoi(apiTimeout); err == nil {
+				data.ApiTimeout = types.Int64Value(int64(timeout))
+			} else {
+				resp.Diagnostics.AddWarning(
+					"Expected OPSLEVEL_API_TIMEOUT to be an int",
+					fmt.Sprintf("OPSLEVEL_API_TIMEOUT was set but not as an int. The default timeout value of %d seconds will be used.", defaultApiTimeout),
+				)
 			}
-			log.Println("[INFO] Initializing OpsLevel client")
+		}
+	}
+	tflog.Debug(ctx, "opslevel client API timeout is set")
 
-			opts := make([]opslevel.Option, 0)
+	opts := []opslevel.Option{
+		opslevel.SetAPIToken(data.ApiToken.ValueString()),
+		opslevel.SetURL(data.ApiUrl.ValueString()),
+		opslevel.SetTimeout(time.Second * time.Duration(data.ApiTimeout.ValueInt64())),
+		opslevel.SetUserAgentExtra(fmt.Sprintf("terraform-provider-%s", p.version)),
+	}
+	client := opslevel.NewGQLClient(opts...)
 
-			opts = append(opts, opslevel.SetAPIToken(token))
-			opts = append(opts, opslevel.SetURL(url))
-			opts = append(opts, opslevel.SetUserAgentExtra(fmt.Sprintf("terraform-provider-%s", version)))
-			opts = append(opts, opslevel.SetTimeout(time.Second*time.Duration(timeout)))
+	tflog.Debug(ctx, "Validating OpsLevel client...")
+	if err := client.Validate(); err != nil {
+		tflog.Error(ctx, fmt.Sprintf("OpsLevel client validation error: %s", err))
+	}
+	tflog.Debug(ctx, "OpsLevel client is valid")
+	tflog.Info(ctx, "OpsLevel client is initialized")
 
-			client := opslevel.NewGQLClient(opts...)
+	resp.DataSourceData = client
+	resp.ResourceData = client
+}
 
-			return client, client.Validate()
-		},
+func (p *OpslevelProvider) Resources(context.Context) []func() resource.Resource {
+	return []func() resource.Resource{
+		NewDomainResource,
+	}
+}
+
+func (p *OpslevelProvider) DataSources(context.Context) []func() datasource.DataSource {
+	return []func() datasource.DataSource{
+		NewDomainDataSource,
 	}
 }
 
-func GetString(d *schema.ResourceData, key string) *string {
-	value := d.Get(key).(string)
-	return &value
+func New(version string) func() provider.Provider {
```

Comment:
should this be `NewProvider` ?

Reply:
It could be 🤷‍♂️ -  I was just following the upgrade examples and example repos